cranelift/codegen/src/isa/x64/inst/mod.rs

```diff
@@ -1649,6 +1710,32 @@ fn x64_map_regs<RUM: RegUsageMapper>(inst: &mut Inst, mapper: &RUM) {
             src.map_uses(mapper);
             map_def(mapper, dst);
         }
+        Inst::XmmRmRImm {
+            ref mut src,
+            ref mut dst,
+            ref op,
+            ..
+        } => {
+            // In certain cases, instructions of this format can convert an XMM register into a
+            // define (e.g. an equality comparison); this extra logic is necessary to inform the
+            // registry allocator of a different register usage.
+            if let RegMem::Reg { reg } = src {
+                if (*op == SseOpcode::Cmppd || *op == SseOpcode::Cmpps) && *reg == dst.to_reg() {
+                    let mut writable_src = Writable::from_reg(*reg);
+                    map_def(mapper, &mut writable_src);
+                    *reg = writable_src.to_reg();
+                    map_def(mapper, dst);
+                } else {
+                    // Otherwise, we map the instruction as usual.
+                    src.map_uses(mapper);
+                    map_mod(mapper, dst);
+                }
+            } else {
+                // TODO this is duplicated because there seems to be no way to join the `if let` and `if`?
```

Comment:
We should be able to do something like:

```
match (a_op_b_is_constant(op), src) {
    (true, RegMem::Reg { reg }) if *reg == dst.to_reg() => { ... }
    _ => { ... }
```

to combine the `if`s. (And probably ignore my half-baked name for the helper...)

Reply:
Ah, nice, let me try that...

Reply:
Hm, `match (..., src)` takes ownership and then I can't do `src.map_uses` later on.